### ISSUE TITLE
Toggling readonly calls "onChange" event

### ISSUE THREAD

**Issue by medzhidov** [codex-team/editor.js#2467]
But it shouldn't. I found that in readonly toggle method we removed all content from it and add it again, but at this time we also have ModificationObserver turned on. So i suggest a solution:

src/components/modules/readonly.ts —> toggle()

```typescript
/**
 * Disable modification observer before changing content
 */
await this.Editor.ModificationsObserver.disable();

await this.Editor.BlockManager.clear();
await this.Editor.Renderer.render(savedBlocks.blocks);

/**
 * Enable modification observer back after content rendered
 */
await this.Editor.ModificationsObserver.enable();
```

**Comment by bhaveshvasnani** [comment]
Hi @neSpecc, Is this issue up for grabs? Could I please work on this to familiarize myself with the project to maybe later contribute to other issues? Thanks!

Also, I tried looking up the code about ModificationsObserver and could you please help me understand if that's just a listener that handles all block change events callbacks? Thus, we wish to suppress those event handle callbacks for this readonly toggle.

**Comment by lijie1129** [comment]
hi, @medzhidov 

Thank you for reporting this issue.

When we discuss and formulate a policy on `readonly`, we need to consider the scenarios of our editor:

* Local standalone scenario
* Multi-person online collaboration scenario


## Local standalone scenario
 
So far, editorjs works very well in the local standalone environment, and the current way of switching readonly mode does not bring much experience problems to the user experience. Similarly, the solution you gave can also solve this problem very well, 👍.

## Multi-person online collaboration scenario 

If editorjs is introduced into the multi-person online collaboration scenario, the current way of switching readonly mode may bring more obstacles to the user experience, and we will face more challenges when trying to optimize it. 

Why is that?

This is because: **_In the multi-person online collaboration scenario, the complexity of dom mutation has increased._**

When readonly switches from disable to enable, two actions are performed:

1. Delete all blocks
2. Then recreate the blocks that were just deleted (the value of `readonly` passed in the constructor of each block will become `true`, and add these built blocks to the editor

The above method will produce two groups of dom mutations:

* The first batch of dom mutations is triggered by deleting all blocks
* The second batch of dom mutations is triggered by adding (the blocks that were just deleted and switched to read-only mode) blocks

For data services, switching readonly should be an atomic action:

* When readonly switches from disable to enable, we usually want to cut off the response of the data layer (avoid unnecessary data persistence requests)
* When readonly switches from enable to disable, we usually want to activate the response of the data layer (allow all data changes to trigger persistence requests)
* When readonly switches, we hope that the data is reliable and complete (there will be no data loss problems, which usually requires putting the action of deleting all blocks and re-adding these blocks into a single transaction)

In summary, when we consider the multi-person online collaboration scenario, we have to take care of those mutations that occur after the dom mutation observer is disconnected, because some of them are likely not triggered by the readonly switch. For example: they may come from the actions triggered by another peer.

For dealing with this kind of scenario, maybe [takeRecords](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver/takeRecords) can help. 😊

**Comment by chf-gh** [comment]
I ran into this issue too, but when I changed readonly to true, the document was already set to not editable behind the scenes, but readonly triggered onchange and deleted all blocks. I just want to change the document property to not be editable without triggering the onchange event. Is there any way?

**Comment by bhaveshvasnani** [comment]
@lijie1129, I have been assigned this issue and am a new contributor to this project. I read your comment and it made a lot of sense to me logically but there are gaps in my understanding when going about the likely fox for this issue. Could you please provide me with a little more context so I could help with this issue?

- Are we looking to handle the single user vs multi-user scenarios differently for the readonly toggle? How do we differentiate between the usage scenario in code?
- And could you please also tell me how would takeRecords would work here inside the toggle flow?

**Comment by lijie1129** [comment]
> @lijie1129, I have been assigned this issue and am a new contributor to this project. I read your comment and it made a lot of sense to me logically but there are gaps in my understanding when going about the likely fox for this issue. Could you please provide me with a little more context so I could help with this issue?
> 
> 
> 
> - Are we looking to handle the single user vs multi-user scenarios differently for the readonly toggle? How do we differentiate between the usage scenario in code?
> 
> - And could you please also tell me how would takeRecords would work here inside the toggle flow?

Hi,@bhaveshvasnani

I'm very glad that you can contribute to the EditorJS. 

I'm currently on holiday, and in a few days, I will prepare relevant information and do my best to provide you with relevant information to solve this problem. 🤝

**Comment by lijie1129** [comment]
hi, @bhaveshvasnani 

The following is the information I can provide to you based on my personal perspective of cognition, I hope it will be helpful to you.

First of all, I think our discussion on `readonly` is to **make EditorJS better support multi-person collaborative editing scenarios.**

# Background information

## What is a single input source editing scenario

From the creation of the document to the final delivery, only one input source can write to the document content at a time.

The change pattern of the document content is divergent, each copy will retain its own version of the content, and there will be multiple different versions of the content after multiple edits of the same original content.

The characteristics of this editing scenario are: 

* **Single input source at a time**
* **Divergent content versions**

Yes, this is the most common editing scenario today.

## What is a multi-person collaborative editing scenario

Multiple input sources are allowed to write to the same document at the same time.

The change pattern of the document content is convergent, the content of all copies needs to be consistent after multiple edits of the same original content.

Everyone can see the changes and awareness information of other editors after editing the content, such as cursor position, input status.

The characteristics of this editing scenario are:

* **Multiple input sources**
* **Convergent content versions**
* **Real-time**

It gives editors a feel similar to the one in the image below:

<img width="638" alt="image" src="https://github.com/codex-team/editor.js/assets/5371749/3b5d63aa-288c-4c4a-9b8e-dea2569c35a0">

If you want to experience the multi-person collaborative editing scenario yourself, you can enter this [link](https://slate-yjs.dev/remote-cursors-overlay). (Remember to open at least two browser windows or tabs).

# The problems brought by multi-person collaborative editing scenarios

Multi-person collaborative editing requires the document content to converge in the end, so in order to achieve this goal, there needs to be a special solution for this. The common solutions currently are:

* [OT(Operational Transformation)](https://en.wikipedia.org/wiki/Operational_transformation)
* [CRDT(Conflict-free replicated data type)](https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type)

No matter which of the two solutions mentioned above we introduce, our goal of achieving multi-person collaborative editing becomes more challenging.

For more than two years, I have been working on the second solution. I use a tool called [YJS](https://github.com/yjs/yjs), which is a CRDT framework.

<img width="480" alt="image" src="https://github.com/codex-team/editor.js/assets/5371749/c12f7a12-7d02-4d44-975d-400652c6efeb">

After introducing YJS, our entire service will roughly become like this:

<img width="919" alt="image" src="https://github.com/codex-team/editor.js/assets/5371749/de991a75-9dd4-4438-aaa1-84e927f53ad4">

We see that we cannot directly use native data formats (such as JSON) to achieve data convergence, but we need to convert native data to YJS format, which requires us to do more work, such as: setting up a two-way binding mechanism between the editor and YJS.

1. The changes in the editing content will trigger the changes in the YJS document, which will trigger the network data transmission and persist to the server (the server can also be regarded as another remote peer)
2. The changes in the content on the remote peer will trigger the network data transmission, the data will reach the local through the network, trigger the local YJS document to change, and thus trigger the editor content update.

# Suggestions on how to better implement readonly

Minimize the cost of editor content changes:

## Reduce unnecessary DOM operations, only do necessary DOM operations

Compared to operating the entire block’s DOM, modifying only one attribute of the block (such as `contenteditable`) costs much less.

## The editor core only makes rules, and is responsible for the necessary supervision, and tries to delegate the specific implementation to the plugins

The editor core’s current specific work can be gradually degraded to default values, giving higher priority to the specific implementation of plugins.

**Comment by lijie1129** [comment]
> * Are we looking to handle the single user vs multi-user scenarios differently for the readonly toggle? How do we differentiate between the usage scenario in code?

hi, @bhaveshvasnani 

I personally suggest that the code logic should be kept simple, and do not make distinctions for different scenarios in the code, find a more general solution that covers multiple scenarios, which is the best.

**Comment by lijie1129** [comment]
> * And could you please also tell me how would takeRecords would work here inside the toggle flow?

hi, @bhaveshvasnani 

As I introduced in the above multi-user collaborative editing scenario, in this scenario, the source of triggering DOM Mutation changes from “only triggered by local editor” to “possibly triggered by local editor” and “possibly triggered by remote peer editor”. 

This increases the possibility that a few DOM Mutation records that should be processed are ignored when disable ModificationsObserver, so if we use `ModificationsObserver.disable();` to reduce the impact of `readonly`, then we need to handle these DOM Mutations.

**Comment by bhaveshvasnani** [comment]
@lijie1129, I genuinely feel that the effort you put into explaining the entire scenario is exceptional and I really appreciate that.

> This increases the possibility that a few DOM Mutation records that should be processed are ignored when disable ModificationsObserver, so if we use ModificationsObserver.disable(); to reduce the impact of readonly, then we need to handle these DOM Mutations.

I think I understood your point here and also totally agree with it that we might miss out on handling other key DOM events when disabled. That's why the original conversation on something like `takeRecords()` was important to be able to correctly process what could not be processed earlier.

> I personally suggest that the code logic should be kept simple, and do not make distinctions for different scenarios in the code, find a more general solution that covers multiple scenarios, which is the best.

Yes, I am currently not sure what a generic solution here could be I would research/think a bit about it and reach back to you.

**Comment by bhaveshvasnani** [comment]
> Suggestions on how to better implement readonly
Minimize the cost of editor content changes:
Reduce unnecessary DOM operations, only do necessary DOM operations
Compared to operating the entire block’s DOM, modifying only one attribute of the block (such as contenteditable) costs much less.
The editor core only makes rules, and is responsible for the necessary supervision, and tries to delegate the specific implementation to the plugins
The editor core’s current specific work can be gradually degraded to default values, giving higher priority to the specific implementation of plugins.

@lijie1129, thanks a lot for your suggestions. I'll keep these in mind and please give me a few days to actually research and identify what could be done here (especially keeping in mind the multi-user scenario). The editor core delegating work to the plugins is currently a black-box for me so I would look up in code to understand where and how it's currently being done at the moment.

**Comment by medzhidov** [comment]
@lijie1129 Thank you a lot fot this comprehensive answer, but I already found that my suggestion is not so good as it can be and  suggested another solution - [make "data-muation-free" more usable](https://github.com/codex-team/editor.js/issues/2205#issuecomment-1703807444). Right now this attribute is useless because we need to add this attribute to EACH element inside content. For example, if we need to include some custom blocks (for example pdf/pptx reader to include embed presentations inside our content by pdf.js) current implementation creates a hell of infinite attributes which we need to control. I suggest one of two ways: add additional attribute to ignore all subchilds, or just change current behaviour

**Comment by medzhidov** [comment]
@lijie1129 So ho we can fix this bug with readonly?

**Comment by chf-gh** [comment]
My temporary solution: use two instances of editor, one readonly and one normal editor, with the front-end displaying that editor via "v-show". Don't use "v-if", or the editor will give you an error.

**Comment by bhaveshvasnani** [comment]
@chf-gh, I am not an expert here but from my perspective having two versions here might lead to more complexity and consistency issues since we have to replicate the same set of commands for both instances of the editor. It can lead to errors on either side creating conflicting versions of the document difficult to maintain a source of truth. Therefore, this may not be best suited for this particular case.

@lijie1129, could better answer this and also comment on whether that temporary solution would gel well with the YJS solution that is being worked on.

**Comment by bhaveshvasnani** [comment]
> Compared to operating the entire block’s DOM, modifying only one attribute of the block (such as contenteditable) costs much less.

@lijie1129 could you please clarify if I get the correct understanding for this one -> does this mean that we only pick up elements that have `contenteditable` set when saving and rendering blocks so we only re-render blocks that are affected by readonly toggle?
Thus, a variation of save functionality that accepts parameters for readonly scenario and picks up the required set of affected blocks could be a direction to look into?

**Comment by lijie1129** [comment]
> @chf-gh, I am not an expert here but from my perspective having two versions here might lead to more complexity and consistency issues since we have to replicate the same set of commands for both instances of the editor. It can lead to errors on either side creating conflicting versions of the document difficult to maintain a source of truth. Therefore, this may not be best suited for this particular case.
> 
> @lijie1129, could better answer this and also comment on whether that temporary solution would gel well with the YJS solution that is being worked on.

hi, @chf-gh 

By preparing a separate editor instance for read-only mode, it indeed introduces more complexity. In the collaborative editing scenario, each document with a different ID maintains a YJS format data model in memory, and the above scheme will cause the YJS format data model to correspond to two editor instances (one for editable mode and another for read-only mode), or two identical YJS format data models correspond to two editor instances with the same ID, which will be more difficult to control. 

This method can be used as an experiment in your private project, but it is not recommended for use in production environments.

**Comment by lijie1129** [comment]
> > Compared to operating the entire block’s DOM, modifying only one attribute of the block (such as contenteditable) costs much less.
> 
> @lijie1129 could you please clarify if I get the correct understanding for this one -> does this mean that we only pick up elements that have `contenteditable` set when saving and rendering blocks so we only re-render blocks that are affected by readonly toggle? Thus, a variation of save functionality that accepts parameters for readonly scenario and picks up the required set of affected blocks could be a direction to look into?

hi, @bhaveshvasnani 

The following is my personal opinion, which can be used as a reference. 

# The specific BlockTool is responsible for switching to read-only mode

The editor core should not do the specific work of switching to read-only mode, these work should be done by each specific blockTool, because only the specific type of blockTool knows how they should switch to read-only mode.

For example: 
* The paragraph type of blockTool may simply remove the `contenteditable` attribute to achieve read-only mode;
* The paragraph type of block and the list type of block have different DOM structures internally, so they have to deal with the contenteditable attribute located in different DOM nodes when switching to read-only mode; 
* The delimiter type of blockTool may not need to do anything, because its content is not editable in the first place;
* The image type of blockTool may switch to read-only mode in a way that is irrelevant to the contenteditable attribute (maybe just hide some menus or block some click events).

Therefore, it is recommended that like the editor requires all blockTools to implement the `save` method, all blockTools are required to implement the `toggleReadonly` method, and the editor core logic is only responsible for calling each blockTool’s toggleReadonly method.

# Read-only mode is isolated from content changes

Switching to read-only mode does not mean that the editor content has changed semantically, read-only mode means that the content editing entry is closed and only reading is allowed.

Therefore, it is recommended that the editor switch to read-only mode and do not trigger the `onChange` event.

I think our current challenge is: how to properly handle the DOM Mutation triggered by each blockTool’s operation on the DOM when switching to read-only mode, which is still something that needs to be considered and explored.

**Comment by bhaveshvasnani** [comment]
@lijie1129, I just wanted to ensure I'm exploring the right direction. So, I wrote a basic boilerplate logic for how we would technically want the Blocks to handle their own implementation of toggling read-only as an override function. I imagined things to be caught at the editor level and then handled over block by block.

I kept the default functionality as not implemented for which we would save the data and rerender the entire block like the current behavior. Could you please guide me if this is in the right direction and then I could look deeper?

Draft PR: https://github.com/codex-team/editor.js/pull/2513

I also wanted some help in understanding how not to trigger an onChange event. I could only see that ModificationsObserver has a disabled flag which could suppress the onChange event. Moreover, I still couldn't understand where and how each type of block would override toggleReadOnly, and would appreciate any pointers on understanding that.

**Comment by lijie1129** [comment]
> @lijie1129, I just wanted to ensure I'm exploring the right direction. So, I wrote a basic boilerplate logic for how we would technically want the Blocks to handle their own implementation of toggling read-only as an override function. I imagined things to be caught at the editor level and then handled over block by block.
> 
> I kept the default functionality as not implemented for which we would save the data and rerender the entire block like the current behavior. Could you please guide me if this is in the right direction and then I could look deeper?
> 
> Draft PR: #2513
> 
> I also wanted some help in understanding how not to trigger an onChange event. I could only see that ModificationsObserver has a disabled flag which could suppress the onChange event. Moreover, I still couldn't understand where and how each type of block would override toggleReadOnly, and would appreciate any pointers on understanding that.

Good job！ @bhaveshvasnani 

Yes, we are on the right direction. 

For blockTools that are not yet compatible with the new protocol (meaning they haven’t implemented the new `toggleReadonly` method), downgrading to the existing logic for backward compatibility is necessary. 

This is important because it’s challenging to ask users to update all their blockTools at once, especially considering that some users may have already implemented custom blockTools.

Regarding better ways to avoid triggering the `onChange` event, I understand that my current knowledge is not much different from yours. It’s an area we both need to explore further. 

In the coming weeks, I might be preoccupied with searching for new job opportunities, but I believe you’ll find a suitable solution to this issue! 😊🚀

**Comment by bhaveshvasnani** [comment]
Thank you so much @lijie1129! I'll continue to look more into this and hopefully come up with a good solution.

I'm also in a similar boat and can understand the commitment that goes into this. All the best for your search and I genuinely hope that you land your dream opportunity at the earliest!!

**Comment by neSpecc** [comment]
Thanks for the discussion. 

I've made a hotfix for that problem. 

1. Blocks won't be re-rendered on initialization in the read-only mode
2. The `onChange()` won't be fired on `readonly.toogle()` API call

We're still using re-rendering for toggling. In the future, the Tools API will get a method for toggling read-only mode on a fly, so we can eliminate this re-rendering
